### ISSUE TITLE
Unnecessary calls to CIS (DXA1.4/SDL Web 8)

### ISSUE THREAD

**Issue by NiclasCedermalm** [sdl/dxa-web-application-java#40]
There is a lot of calls to retrieve publication mappings from the CIS layer for all URLs (HTML and all assets) when rendering DXA pages. For example the standard DXA home page will result in 34(!) requests to the CIS layer asking for publication mappings each time. See below trace from my CIS log:

`016-04-20 09:46:11,482 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:11,482 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:12,568 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:12,568 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:14,077 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:14,077 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:15,654 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:15,654 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:44,304 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:44,304 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:48,021 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:48,021 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/index
2016-04-20 09:46:48,386 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/system/v1.4/assets/scripts/header.js
2016-04-20 09:46:48,386 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/system/v1.4/assets/scripts/header.js
2016-04-20 09:46:48,386 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/system/v1.4/assets/css/main.css
2016-04-20 09:46:48,386 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/system/v1.4/assets/css/main.css
2016-04-20 09:46:48,389 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/system/v1.4/assets/scripts/main.js
2016-04-20 09:46:48,389 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/system/v1.4/assets/scripts/main.js
2016-04-20 09:46:48,394 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/media/logo_tcm32-709.png
2016-04-20 09:46:48,394 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/media/logo_tcm32-709.png
2016-04-20 09:46:48,427 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/system/v1.4/assets/scripts/xpm.js
2016-04-20 09:46:48,427 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/system/v1.4/assets/scripts/xpm.js
2016-04-20 09:46:48,573 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/media/company-news-placeholder_tcm32-688_w1024_n.png
2016-04-20 09:46:48,573 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/media/company-news-placeholder_tcm32-688_w1024_n.png
2016-04-20 09:46:48,638 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/media/bulls-eye_tcm32-693_w1024_h311_n.JPG
2016-04-20 09:46:48,638 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/media/bulls-eye_tcm32-693_w1024_h311_n.JPG
2016-04-20 09:46:48,641 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/media/map_tcm32-695_w1024_h311_n.jpg
2016-04-20 09:46:48,641 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/media/map_tcm32-695_w1024_h311_n.jpg
2016-04-20 09:46:48,641 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/media/ballon-burner_tcm32-699_w1024_h311_n.jpg
2016-04-20 09:46:48,641 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/media/ballon-burner_tcm32-699_w1024_h311_n.jpg
2016-04-20 09:46:48,667 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/system/v1.4/assets/fonts/fontawesome-webfont.woff
2016-04-20 09:46:48,667 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/system/v1.4/assets/fonts/fontawesome-webfont.woff
2016-04-20 09:46:49,122 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/favicon.ico
2016-04-20 09:46:49,122 INFO  GetPublicationMappingsFunction - Started processing function call for url : http://preview:8080/favicon.ico`

As you see for each URL the request is sent twice. The actual code in com.sdl.webapp.tridion.TridionLocalizationResolver is the following:

`'PublicationMapping publicationMapping = dynamicMappingsRetriever.getPublicationMapping(url);`

So this must be possible to cache to avoid all those calls to the CIS layer.

**Comment by rpannekoek** [comment]
The CIL can/will actually cache the information for indentical URLs if you have CIL caching enabled. However, on a Staging site, CIL caching is typically disabled to preview caching issues in Session Preview.  This is one of the reasons why we use different configurations for Staging and Live.

**Comment by NiclasCedermalm** [comment]
True. But do we really want 34 CIL requests per page request even on a staging site? We try to optimize the XPM performance, but introducing a very talky interaction with CIL will not help. The publication mapping info is not something that is changed very often, so we could use some other kind of caching strategy/configuration here.

**Comment by rpannekoek** [comment]
Unfortunately, the current CIL API doesn't really support caching the mappings in DXA; the "problem" is that the first part of the path may be relevant for the mapping (best-match algorithm). In order to do a proper best-match implementation, you have to know all possible mappings which can currently not be obtained using the CIL API.  Indeed makes sense to optimize this in the future, but it's not "low-hanging fruit".

**Comment by NiclasCedermalm** [comment]
Ok. Add this as an enhancement then. I'll do some temporary workarounds for this in my own fork to start with.

**Comment by bengineer-** [comment]
Are any of these calls to the page linking API? If so, you could use the new Batch Linking API to reduce repetitive calls to the Linking API and send in one request instead.

**Comment by NiclasCedermalm** [comment]
Yes, some calls are done for resolving component links. For the DXA default home page it is 10 requests for that towards CIL. So I guess this is another improvement we can add to the list.

**Comment by rpannekoek** [comment]
Awaiting your Pull Request ;-)

**Comment by NiclasCedermalm** [comment]
Hahaha :-)

I did the following simple extension in my webapp (did not even have to modify the DXA framework code :-) ) to get around the chatty GetPublicationMappings calls:

`@Component
@Primary
public class CachedTridionLocalizationResolver extends TridionLocalizationResolver {

    private Cache<String, PublicationMappingData> cachedPublicationMappings = CacheBuilder.newBuilder()
            .maximumSize(100).expireAfterWrite(5, TimeUnit.MINUTES).build();

    @Override
    protected PublicationMappingData getPublicationMappingData(String url) throws PublicationMappingNotFoundException {

        PublicationMappingData publicationMapping = cachedPublicationMappings.getIfPresent(url);
        if ( publicationMapping == null ) {
            publicationMapping = super.getPublicationMappingData(url);
            this.cachedPublicationMappings.put(url, publicationMapping);
        }
        return publicationMapping;
    }
}
`

**Comment by majiccode** [comment]
I've already raised an enhancement to move to the new batch linking API. This isn't related to that though.

**Comment by rpannekoek** [comment]
TSI-1689 (internal issue ID for tracking purposes)

**Comment by rpannekoek** [comment]
In DXA 2.0 we are introducing a DXA Model Service which is intended to be deployed next to the Content Service.  The main purpose of this new Component is to minimize the traffic between Web Application on CIS backend: only one roundtrip is needed to get all the data needed to render a Page Model (incl. Include Pages, DCPs, linked/embedded Entity Models and resolved links).
